[#3159] Support Apache Avro

zambrovski opened the pull request:
This pull request resolves #3159.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AxonFramework/AxonFramework?pullRequest=3187) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/AxonFramework/AxonFramework?pullRequest=3187) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: jangalinski<br/>:x: zambrovski<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AxonFramework/AxonFramework?pullRequest=3187) it.</sub>

zambrovski commented:
Not ready for the final review yet.

zambrovski commented on integrationtests/pom.xml:
```diff
@@ -242,11 +242,6 @@
             <scope>test</scope>
         </dependency>
 
-        <dependency>
```
resolved duplication warning in the build

smcvb commented on spring/src/test/java/org/axonframework/spring/serialization/avro/test1/ComplexObject.java:
```diff
@@ -0,0 +1,477 @@
+/**
```
Same class-level JavaDoc and copyright question as with the previous `ComplexObject`.

smcvb commented on spring/src/main/java/org/axonframework/spring/serialization/avro/ClasspathAvroSchemaLoader.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Classpath Avro schema loader.
+ */
+public interface ClasspathAvroSchemaLoader {
```
Ah, this is your own custom loader. Then I think my comment on the auto-config class is invalid.

smcvb commented on spring/src/main/java/org/axonframework/spring/serialization/avro/AvroSchemaScan.java:
```diff
@@ -0,0 +1,73 @@
+@Target(ElementType.TYPE)
+@Documented
+@Inherited
+public @interface AvroSchemaScan {
```
Cool stuff! :ok_hand:

smcvb commented on spring/src/main/java/org/axonframework/spring/serialization/avro/SpecificRecordBaseClasspathAvroSchemaLoader.java:
```diff
@@ -0,0 +1,79 @@
+
+    private static Schema getClassSchema(Class<SpecificRecordBase> clazz)
+            throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        return (Schema) clazz.getMethod("getClassSchema").invoke(null); // invoke static method
```
So, this approach might fail if, for whatever reason, the scanned-for packages contain classes that have a `static getClassSchema` method that does not align with what we expect, right?

jangalinski replied:
This is an implicit contract: All classes generated for java inherit from SpecificRecordBase and have default static methods generated. So you could theoretically write your own class breaking this contract, but is is highly unlikely and discouraged and then it could break, but that's ok. We do try/catch around this call and would just not load such misused classes.

smcvb reviewed (changes requested):
There are some things I haven't completely reviewed yet, but I think I have enough comments. The amount is the reason for me to request changes, by the way.

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,247 @@
+        if (serializerStrategy.isPresent()) {
+
+            // with upcasting:
+            // GenericRecord -> T
```
GenericRecord is an Avro representation of a map. You can add/remove/modify fields of the record on it. So the upcasting strategy is bytes -> GenericRecord, modify GenericRecord and then GenericRecord -> Object (T)...

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializerStrategy.java:
```diff
@@ -0,0 +1,70 @@
+                                             @Nonnull Class<T> type);
+
+    /**
+     * Deserializes from Apache Avro generic record (intermediate representation).
```
exactly

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroUtil.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    /**
+     * Reads fingerprint from byte array.
```
exactly. This is a "hash" of the schema with a guarantee, that it changes if the schema changes.

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/ByteArrayToGenericRecordConverter.java:
```diff
@@ -0,0 +1,66 @@
+import java.io.IOException;
+
+/**
+ * Content type converter between single-object-encoded bytes and Avro generic record.
```
Why should you want one? if you serialize, you go from T->bytes directly. If you deserialize, you either go bytes->T (no upcasting), or bytes->GenericRecord->T (with upcasting). We can provide one, but there is no clear reason for me why...

zambrovski replied:
@smcvb Should we supply one? I don't understand why...

vab2048 replied:
It could potentially be useful for those storing byte[] in the DBs? Or for having it available for certain tests.

smcvb replied:
Thanks for including it regardless, @zambrovski :pray:

zambrovski commented on spring-boot-autoconfigure/src/main/java/org/axonframework/springboot/autoconfig/AvroSerializerAutoConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+     * @return schema store instance.
+     */
+    @Bean("defaultAxonSchemaStore")
+    // TODO we need to handle missing bean condition too in order to make it replaceable.
```
I thought by default declared conditions are logically combined with OR (not AND).

smcvb replied:
I think this point still should be covered before we can go ahead and merge this, correct?

smcvb replied:
I'll round this off in a separate commit, to unblock this already immense effort for a new serializer.

zambrovski commented on spring-boot-autoconfigure/src/main/java/org/axonframework/springboot/autoconfig/AvroSerializerAutoConfiguration.java:
```diff
@@ -0,0 +1,117 @@
+    @Bean("defaultAxonSchemaStore")
+    // TODO we need to handle missing bean condition too in order to make it replaceable.
+    @Conditional({AvroConfiguredCondition.class})
+    public SchemaStore defaultAxonSchemaStore(BeanFactory beanFactory, List<ClasspathAvroSchemaLoader> schemaLoader) {
```
`ClasspathAvroSchemaLoader` is an interface. The implementation is working for the currently supported loader, but for Kotlin, we will provide a different implementation.

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,247 @@
+                .findFirst();
+
+        if (serializerStrategy.isPresent()) {
+            if (byte[].class.equals(expectedRepresentation)) {
```
`byte[]` and `GenericRecord`
What other expected representations should we consider?

zambrovski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,247 @@
+    public <T> SerializedObject<T> serialize(Object object, @Nonnull Class<T> expectedRepresentation) {
+        // assume: expectedRepresentation is byte[] (or possibly String), nothing else.
+        // TODO: Question is this correct? What is the contract of null handling?
+        Objects.requireNonNull(object, "Can't serialize a null object");
```
Valid. Since there is a inconsistency in other serializers.

What is your expectation?

We can of course support `null` by returning an emptySerializedObject...

zambrovski replied:
I mean there is an asymmetry between serialize/deserialize

zambrovski replied:
Json serializer test creates a null object and serializes it. The Jackson JSON representation of this null is "null". I believe this is not valid JSON, even if Jackson does it this way.

zambrovski replied:
We will delegate to serializerDelegate :-)

filpano commented:
Hi - first off, great stuff. Love seeing this get more traction for interoperability.
Coming from a Kafka side of things, I have had to additionally configure the following properties on the consumption side:

- `specific.avro.reader` - Map to the `SpecificRecord` Types as opposed to `GenericRecord`
- `avro.use.logical.type.converters` - Try to honor the `logicalType` definition in the schema (i.e. `BigDecimal` type instead of `ByteBuffer`), and
- `avro.remove.java.properties` - Removes Java-specific properties from the generated schemas (i.e. the `$SCHEMA` field) such as `avro.java.string` (which the Avro Maven Plugin likes to add, see the note [here](https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/serdes-avro.html#avro-serializer) at the bottom of this section).

I flew over the PR but did not notice too much in the way of configuration options regarding the above three concerns (it is possible I have overlooked this). 

I think it would be fine if Axon provided an opinionated approach (i.e. setting the last two properties to `true`), but this should probably be explicitly documented. I know for a fact that `avro.remove.java.properties: false` causes interoperability issues with e.g. Confluent Schema Registry.

Hopefully this comment can save a bit of pain down the road. :)

zambrovski commented:
> Hi - first off, great stuff. Love seeing this get more traction for interoperability. Coming from a Kafka side of things, I have had to additionally configure the following properties on the consumption side:
> 
> * `specific.avro.reader`
>   
>   * Map to the `SpecificRecord` Types as opposed to `GenericRecord`
> * `avro.use.logical.type.converters`
>   
>   * Try to honor the `logicalType` definition in the schema (i.e. `BigDecimal` type instead of `ByteBuffer`), and
> * `avro.remove.java.properties`
>   
>   * Removes Java-specific properties from the generated schemas (i.e. the `$SCHEMA` field) such as `avro.java.string` (which the Avro Maven Plugin likes to add, see the note [here](https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/serdes-avro.html#avro-serializer) at the bottom of this section).
> 
> I flew over the PR but did not notice too much in the way of configuration options regarding the above three concerns (it is possible I have overlooked this).
> 
> I think it would be fine if Axon provided an opinionated approach (i.e. setting the last two properties to `true`), but this should probably be explicitly documented. I know for a fact that `avro.remove.java.properties: false` causes interoperability issues with e.g. Confluent Schema Registry.
> 
> Hopefully this comment can save a bit of pain down the road. :)

Interesting. Where do you configure this properties, and what stack are you using? Is it confluent? 

Current implementation assumes that your classes have been generated by Apache Maven Avro Plugin. What tool are you using to generate Java classes out of Avro schema?

Cheers,

Simon

filpano commented:
>Interesting. Where do you configure this properties, and what stack are you using? Is it confluent?

We're using a Confluent Schema Registry together with Kafka for schema management. Our use case goes like: producer registers schemas to Schema Registry -> Consumers fetch schemas -> Consumers generate class files from schemas.

The properties I mentioned are all consumer properties that interact with the Confluent `KafkaAvroDeserializer`:

```yaml
# Only relevant when using io.confluent.kafka.serializers.KafkaAvroDeserializer
kafka.consumer.properties.avro.use.logical.type.converters: true
kafka.consumer.properties.avro.remove.java.properties: true
kafka.consumer.properties.specific.avro.reader: true
```

These may seem similar to the avro-maven-plugin settings - because they are essentially the same settings, configurable at deserialization time. See [KafkaAvroDeserializerConfig](https://github.com/confluentinc/schema-registry/blob/master/avro-serializer/src/main/java/io/confluent/kafka/serializers/KafkaAvroDeserializerConfig.java#L26).

>Current implementation assumes that your classes have been generated by Apache Maven Avro Plugin. What tool are you using to generate Java classes out of Avro schema?

The resulting Java-specific string was generated by both the `avro-maven-plugin` (as of `1.11.0`, by now a little outdated) as well as the unofficial [Avro Gradle Plugin](https://github.com/davidmc24/gradle-avro-plugin) (which has since been archived and donated to Apache, so that's kind of in-between right now).

The problem lies within the `$SCHEMA` of the generated classes themselves, i.e.:

```json
...
{
  "name":"myStringProperty",
  "type":{
    "type":"string",
    "avro.java.string":"String"
  }
}
...
```

The cause is the following Avro bug: https://issues.apache.org/jira/browse/AVRO-2838. 

Due to this, I have to pass the `consumer.properties.avro.remove.java.properties: true` to my Kafka Consumers which are Schema Registry aware, due to the schema being semantically changed by the generator (the types are not compatible as-is). I am not aware as to whether other Schema Registries have similar handling. 

I suppose the required handling here (if any) depends on where the Schemas are managed. I do not think it makes sense to support something like Confluent Schema Registry directly as this requires Kafka (at the time of this writing).

What it boils down to is that this probably should be at least noted somehow in the documentation under "Requirements" or something similar, otherwise users may get a nasty surprise when they try to connect different systems using a centralized schema management (which, as far as I can tell, is outside of the scope of this issue/PR).

>In doing so, the logical type resolution is performed by the SPI-like detection of the original Avro Jva Stack and the $SCHEMA is part of the SpecificRecordBase class.

If the `$SCHEMA` field is parsed at all, the above issue may be encountered. I don't see any way around this other than explicitly handling the Avro generator bug (which is what Confluent has essentially done), but I'm not super deeply versed in the Avro generator specifics.

If I can help out with further information, please let me know.

zambrovski commented:
I followed your link. I think we can describe this, but:

1. The schema included into the class is full pairwise-compatible to the written "original" using the `SchemaCompatilibty.checkReaderWriterCompatibility`.
2. The fact that Confluence is determining the schema as incompatible in their stack, is probably due to the fact that Confluent is not paying much attention to the original Avro Specification. There ist is defined how the schemas are compared: https://avro.apache.org/docs/1.12.0/specification/#schema-resolution. In fact it is absolutely ok to add any number of additional attributes (which are not reserved by Avro specification) on any node (this makes Avro extensible format), and those should not break the compatibility. Having said that - sorry if Confluent ignores that fact and breaks because the Code generator includes the legally modified schema into a static field. 

With the arguments above, I don't think it is a problem of Avro or problem of this serializer. At the end, we a static method `SpecificRecordBased#getClassSchema`. Even if the proposal of the referenced issue will be implemented, we will still get the correct schema, and the schema is compatible with the hand-written by the user.

My two cents on that..

Cheers,

Simon

smcvb reviewed (commented):
Sorry for the long wait on my re-review; quite a lot happened the last month that required my focus. But, here it is, my second review. Still a couple of pointers, but nothing really major.

I am not chosen to approve or request changes _yet_ though, as I have two unknowns I would like to discuss:

1. Documentation - so, a section / paragraph about this in the [Serialization page](https://docs.axoniq.io/axon-framework-reference/development/serialization/).
2. Upcasting - how would this work right now?

The documentation pointer is a request, but not a necessity. We have quite recently moved the documentation to be included in Axon Framework itself. Hence, a pull request can contain the documentation for fixes, enhancements, and additions. **However**, I don't want to impose this as a requirement on your plate (hence why I mark this as an unknown), given the scope of what you're delivering. Nonetheless, let me know if you're up to the task.

And the second point, upcasting. I am aware that the `AvroSerializerStrategy` has an upcasting flow, through the `AvroSerializerStrategy#deserializeFromGenericRecord` operation. However, I haven't found where this is being used, if at all. So, if it is used, could you point me to that place? And, if it is not used, could you give a run down **how** it would work?

Thanks again for all the effort here, @zambrovski and @jangalinski! :raised_hands:

jangalinski commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,373 @@
+     * <p>
+     * The following fields are mandatory, to create the Serializer:
+     * <ul>
+     *     <li>{@link Builder#revisionResolver(RevisionResolver)}, to resolve revisions of events</li>
```
Added javadoc.

smcvb replied:
Ah, I see it on the class-level JavaDoc, thanks! But, could you also add it to this `builder()` method and to the `Builder` class itself perhaps?

jangalinski commented on messaging/src/main/java/org/axonframework/serialization/avro/SpecificRecordBaseSerializerStrategy.java:
```diff
@@ -0,0 +1,183 @@
+        SpecificData readerSpecificData = SpecificData.getForClass(specificRecordBaseClass);
+        Schema readerSchema = AvroUtil.getClassSchemaChecked(specificRecordBaseClass);
+
+        if (this.avroSerializerStrategyConfig.performAvroCompatibilityCheck()) {
```
we will investigate ... do not like the idea of static caching

smcvb replied:
I think this has been resolved nicely! Hence, I'll go ahead and resolve this comment.

jangalinski commented on spring-boot-autoconfigure/src/main/java/org/axonframework/springboot/autoconfig/AvroSerializerAutoConfiguration.java:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        @SuppressWarnings("unused")
+        // TODO: Question, the use case of events and messages makes sense for Avro, but general is artificial?
```
We will check how this behaves in an example app ... but basically tend to removal.

zambrovski replied:
We performed the check today and though about this feature. The result is:

1. We removed the condition to auto-configure AvroSerializer if it is used as General.
2. We switched the fallback serializer of AvroSerializer from "opininated" Jackson to the configured General one. By doing so, we give the control back to the user.
3. We now forbid to configure AvroSerializer as General by using auto-configurations, because otherwise, the fallback serializer can't be used. It makes no sense in any case to do so, but now, the auto-configuration is more robust.

zambrovski commented:
@smcvb we addressed most of your comments. Please resolve all marked with :+1:

The only one missing is the issue about caching incompatibilities. We are still on it.
The documentation is has to be added still, but this is another repository.

smcvb commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,389 @@
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> SerializedObject<T> serialize(Object object, @Nonnull Class<T> expectedRepresentation) {
+        // assume: expectedRepresentation is byte[] (or possibly String), nothing else.
+        // delegate null handling to serializerDelegate, since Avro can't serialize null.
+        if (object != null) {
+            Optional<AvroSerializerStrategy> serializerStrategy = serializerStrategies
+                    .stream()
+                    .filter(it -> it.test(object.getClass()))
+                    .findFirst();
+
+            // TODO smcvb I only ever use String from the other serializers. Then again, their format is human readable, which isn't the case for Avro, of course.
+            // TODO So, not really an enforcement from my end, just trying to understand this serializer.
+            if (serializerStrategy.isPresent()) {
+                if (byte[].class.equals(expectedRepresentation)) {
+                    return (SerializedObject<T>) serializerStrategy.get().serializeToSingleObjectEncoded(object);
+                }
+            }
+        }
```
To be frank, I am not certain what the status is here when the given `expectedRepresentation` is of type `String`. As that something we can support at all? If not, I think we can preemptively check on `String` and if it is, delegate to the `serializerDelegate`.

zambrovski replied:
Why should we check? Currently, the implementation will fall through to delegate anyways. I removed the FIXME/TODO marker.

smcvb replied:
That sounds fine to me indeed. Thanks for elaborating again, @zambrovski.

smcvb commented on messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,373 @@
+    }
+
+    /**
+     * Builder to set up Avro Serializer.
+     * <p>
+     * The following fields are mandatory, to create the Serializer:
+     * <ul>
+     *     <li>{@link Builder#revisionResolver(RevisionResolver)}, to resolve revisions of events</li>
+     *     <li>{@link Builder#schemaStore(SchemaStore)}, to lookup schemas by their fingerprints</li>
+     *     <li>{@link Builder#serializerDelegate(Serializer)}, to deserialize all non-Avro artifacts</li>
+     * </ul>
+     * </p>
```
As stated on the `builder()` method again, I think that the `Builder` its class-level JavaDoc should specify that an `AvroSerializerStrategy` is a hard requirement. :-)

smcvb commented on messaging/src/main/java/org/axonframework/serialization/avro/DefaultSchemaIncompatibilityChecker.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * Retrieves a list of incompatibilities cased so far.
```
```suggestion
     * Retrieves a list of incompatibilities cached so far.
```

smcvb commented on messaging/src/main/java/org/axonframework/serialization/avro/DefaultSchemaIncompatibilityChecker.java:
```diff
@@ -0,0 +1,69 @@
+     * Visible for testing.
+     * @return copy of immutability map.
+     */
+    Map<Pair<Long, Long>, List<SchemaCompatibility.Incompatibility>> getIncompatibilitiesCache() {
```
I assume that the `getIncompatibilitiesCache` and `clear` methods are package private on purpose?

zambrovski replied:
yes. this is only for testing purposes visible.

smcvb replied:
Cool, thanks for sharing.

smcvb commented on spring-boot-autoconfigure/src/main/java/org/axonframework/springboot/autoconfig/AvroSerializerAutoConfiguration.java:
```diff
@@ -0,0 +1,114 @@
+
+    /**
+     * An {@link AnyNestedCondition} implementation, to support the following use cases:
+     * <ul>
+     *     <li>The {@code general} serializer property is set to {@code avro}</li>
+     *     <li>The {@code messages} serializer property is set to {@code avro}</li>
+     *     <li>The {@code events} serializer property is set to {@code avro}</li>
+     * </ul>
```
```suggestion
     * <ul>
     *     <li>The {@code messages} serializer property is set to {@code avro}</li>
     *     <li>The {@code events} serializer property is set to {@code avro}</li>
     * </ul>
```

smcvb commented on spring/src/main/java/org/axonframework/spring/serialization/avro/SpecificRecordBaseClasspathAvroSchemaLoader.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final ClassPathScanningCandidateComponentProvider candidateProvider;
+
+    public SpecificRecordBaseClasspathAvroSchemaLoader(ResourceLoader resourceLoader) {
```
This public constructor is missing some JavaDoc.

smcvb reviewed (commented):
Some small fry comments remaining! Thanks for taking the time for another pass after my relatively long PR at the end of 2024, @zambrovski and @jangalinski :pray: 

I was off myself for a week and we had a company week the previous one; hence why you had to wait a little.
However, I would like to get this done ASAP. Assuming that suits your schedule. I think it would be good to do a 4.11.0 release of Axon Framework "soonish" and this `AvroSerializer` belongs in there!

smcvb reviewed (approved):
My concerns have been addressed, hence I'm approving this pull request.
